src/lib/measure.c

```diff
@@ -0,0 +1,152 @@
+{
+	struct measure_context context;
+	int num_steps = 1 << steps_log2;
+
+	memset_s(&context, sizeof(struct measure_context), 0,
+		 sizeof(struct measure_context));
+
+	context.avg_shift = steps_log2;
+	context.steps = num_steps;
+	context.samples =
+		(struct measure_sample *)rzalloc(RZONE_RUNTIME,
+						 SOF_MEM_CAPS_RAM,
+						 num_steps *
+						 sizeof(struct measure_sample));
+
+	return context;
+}
+
+void measure_free(struct measure_context *context)
+{
+	rfree(context->samples);
+	context->samples = NULL;
+}
+
+static inline void measure_update_min_stat(struct measure_context *context)
+{
+	struct measure_sample *sample = &context->samples[context->cur_step];
+
+	if (sample->time_delta <
+	    context->stats[MEASURE_STAT_MIN].time_delta)
+		context->stats[MEASURE_STAT_MIN].time_delta =
+			sample->time_delta;
+
+	if (sample->ccount_delta <
+	    context->stats[MEASURE_STAT_MIN].ccount_delta)
+		context->stats[MEASURE_STAT_MIN].ccount_delta =
+			sample->ccount_delta;
+}
+
+static inline void measure_update_max_stat(struct measure_context *context)
+{
+	struct measure_sample *sample = &context->samples[context->cur_step];
+
+	if (sample->time_delta >
+	    context->stats[MEASURE_STAT_MAX].time_delta)
+		context->stats[MEASURE_STAT_MAX].time_delta =
+			sample->time_delta;
+
+	if (sample->ccount_delta >
+	    context->stats[MEASURE_STAT_MAX].ccount_delta)
+		context->stats[MEASURE_STAT_MAX].ccount_delta =
+			sample->ccount_delta;
+}
+
+static inline void measure_update_avg_stat(struct measure_context *context)
+{
+	int avg_step;
+
+	context->stats[MEASURE_STAT_AVG].ccount_delta = 0;
+	context->stats[MEASURE_STAT_AVG].time_delta = 0;
+
+	for (avg_step = 0; avg_step < context->steps; ++avg_step) {
+		context->stats[MEASURE_STAT_AVG].ccount_delta +=
+			context->samples[avg_step].ccount_delta;
+		context->stats[MEASURE_STAT_AVG].time_delta +=
+			context->samples[avg_step].time_delta;
+	}
+	context->stats[MEASURE_STAT_AVG].ccount_delta >>=
+		context->avg_shift;
+	context->stats[MEASURE_STAT_AVG].time_delta >>=
+		context->avg_shift;
+}
+
+static inline void measure_update_stats(struct measure_context *context)
+{
+	measure_update_min_stat(context);
+	measure_update_avg_stat(context);
+	measure_update_max_stat(context);
+}
+
+void measure_begin(struct measure_context *context)
+{
+	context->begin_time = platform_timer_get(platform_timer);
+
+	/* TODO: Handle 32-bit overflow of the CCOUNT register */
+	context->begin_ccount = xthal_get_ccount();
+}
+
+void measure_end(struct measure_context *context)
+{
+	struct measure_sample *sample = &context->samples[context->cur_step];
+	uint64_t cur_time = platform_timer_get(platform_timer);
+
+	/* TODO: Handle 32-bit overflow of the CCOUNT register */
+	uint64_t cur_ccount = xthal_get_ccount();
+
+	sample->time_delta = cur_time - context->begin_time;
+	sample->ccount_delta = cur_ccount - context->begin_ccount;
+
+	measure_update_stats(context);
```

Comment:
This one calls a lot of functions where sample \@cur_step is obtained again and again. If you pass it by ptr (since already stored here in automatic variable), it could help the compiler to optimize across other inlined functions (I'm not sure if it is already smart enough to detect the address computation would result in the same value in the next inlined piece, perhaps yes ;).